src/jit/importer.cpp

```diff
@@ -4390,6 +4394,311 @@ GenTree* Compiler::impBaseIntrinsic(NamedIntrinsic        intrinsic,
+                    valueOp = gtNewSimdHWIntrinsicNode(TYP_SIMD16, valueOp, NI_Base_Vector128_CreateScalarUnsafe,
+                                                       TYP_FLOAT, 16);
+                    immNode->AsIntCon()->SetIconValue(imm8 * 16);
+                    __fallthrough;
+
+                case TYP_BYTE:
+                case TYP_UBYTE:
+                case TYP_INT:
+                case TYP_UINT:
+                    retNode =
+                        gtNewSimdHWIntrinsicNode(TYP_SIMD16, vectorOp, valueOp, immNode, NI_SSE41_Insert, baseType, 16);
+                    break;
+
+                case TYP_SHORT:
+                case TYP_USHORT:
+                    retNode =
+                        gtNewSimdHWIntrinsicNode(TYP_SIMD16, vectorOp, valueOp, immNode, NI_SSE2_Insert, baseType, 16);
+                    break;
+
+                case TYP_DOUBLE:
+                {
+                    assert(imm8 == 0 || imm8 == 1);
+                    valueOp = gtNewSimdHWIntrinsicNode(TYP_SIMD16, valueOp, NI_Base_Vector128_CreateScalarUnsafe,
+                                                       TYP_DOUBLE, 16);
+                    NamedIntrinsic in = (imm8 == 0) ? NI_SSE2_MoveScalar : NI_SSE2_UnpackLow;
+                    retNode           = gtNewSimdHWIntrinsicNode(TYP_SIMD16, vectorOp, valueOp, in, TYP_DOUBLE, 16);
+                    break;
+                }
+
+                default:
+                    unreached();
+                    break;
+            }
+
+            if (simdSize == 32)
+            {
+                assert(clonedVectorOp);
+                int upperOrLower = (cachedImm8 >= count / 2) ? 1 : 0;
+                retNode = gtNewSimdHWIntrinsicNode(retType, clonedVectorOp, retNode, gtNewIconNode(upperOrLower),
+                                                   NI_AVX_InsertVector128, baseType, simdSize);
+            }
+
+            break;
+        }
+
+        case NI_Base_Vector256_GetElement:
+        {
+            if (!compSupports(InstructionSet_AVX))
+            {
+                // Using software fallback if JIT/hardware don't support AVX instructions and YMM registers
+                return nullptr;
+            }
+            __fallthrough;
+        }
+        case NI_Base_Vector128_GetElement:
+        {
+            assert(sig->numArgs == 1);
+            GenTree* indexOp = impStackTop().val;
+            if (!compSupports(InstructionSet_SSE2) || !varTypeIsArithmetic(baseType) || !indexOp->OperIsConst())
+            {
+                // Using software fallback if
+                // 1. JIT/hardware don't support SSE2 instructions
+                // 2. baseType is not a numeric type (throw execptions)
+                // 3. index is not a constant
+                return nullptr;
+            }
+
+            switch (baseType)
+            {
+                // Using software fallback if baseType is not supported by hardware
+                case TYP_BYTE:
```

Comment:
`BYTE` and `UBYTE` should be supported by `SSSE3`
`INT` and `UINT` should be supported by `SSE2` (I imagine you could emulate `LONG`/ULONG` with `PSHUFD` as well)

Reply:
Actually, you can emulate all the types on SSE2, but I do not think it worthy to make such complex (at least in the first step). I prefer to wait a bit to see real scenarios and improve them if indeed necessary.

Reply:
I don't think that emulating is necessarily worthwhile, but when we have explicit instructions that would make it trivial to support, it is likely worth doing so.

Reply:
Likely worth noting, I think most users likely have SSE4.1 support and this is probably not that important for the JIT scenario.

However, it is likely worthwhile for the AOT/Crossgen scenario where SSE2 is the default target.

Reply:
> Actually, you can emulate all the types on SSE2,
> I don't think that emulating is necessarily worthwhile,

I added code in this PR to emulate `Vector128<float/double>.Get/WithElement` for only SSE2 supported hardware. 

In my opinion, for other base-types, we need to unify the hardware accelerating bar. If we decide to optimize the helpers blow SSE4.1, then all the helpers on a base-type should have a unified bar (SSE2? SSE3?). Since the helpers are usually used togather, if one helper is not optimized, other helpers' optimization would be useless.

Will open an issue to track.

Reply:
The helpers that I have implemented so far are all using intrinsics available in the earliest ISA available, where possible. For example, `Create(byte)`  is accelerated even on SSE2 by using `UnpackLow`: https://source.dot.net/#System.Private.CoreLib/shared/System/Runtime/Intrinsics/Vector128.cs,30

These are all similar to optimizations that the native compilers do for the corresponding native intrinsics.

Reply:
As with any of the cases, we should either show that it is an existing optimization being used or that benchmarks show it is worthwhile.

I imagine there are some cases where we could "accelerate" by using SSE2, but where the corresponding code would actually not be faster than just storing and accessing the appropriate element.

Reply:
Yes, so I am considering to emulate all the base types on SSE2, but that is complex and need more perf investigations. So, I suggest to do it in the next steps.

Reply:
Right, but I was just saying that we should likely do the ones that would be clear wins (since they are directly supported by a single instruction and we already know that the `shuffle` mechanism is beneficial from existing compilers/libraries doing the same).

Reply:
That is, using `PSHUFD` for `INT/UINT` on SSE2 and `PSHUFB` for `BYTE/SBYTE` on SSSE3

Reply:
> `PSHUFB` for `BYTE/SBYTE` on SSSE3

This seems not worthwhile if we have codegen bottom-line on SSE2 and better instructions on SSE4.1.

Reply:
I don't agree. We have already implemented all the relevant code paths elsewhere, as do other compilers. We should probably just add the code-path.